caffe2/contrib/tensorrt/tensorrt_tranformer.cc

```diff
@@ -455,20 +413,69 @@ void TensorRTTransformer::ClusterToTrtOp(
   model->mutable_graph()->clear_initializer();
 }
 
+CaffeMap<std::string, TensorShape> TensorRTTransformer::SsaRewriteAndMapIO(
+    Workspace* ws,
+    NetDef* pred_net,
+    const std::unordered_map<std::string, TensorShape>& input_shape_hints) {
+  input_mapping_ = onnx::SsaRewrite(nullptr, pred_net);
+  std::unordered_map<std::string, std::string> input_reverse_mapping;
+  std::vector<std::string> external_inputs;
+  for (const auto kv : input_mapping_) {
+    input_reverse_mapping.emplace(kv.second, kv.first);
+    if (!ws->HasBlob(kv.second)) {
+      external_inputs.emplace_back(kv.first);
+    }
+  }
+  for (const auto& i: external_inputs) {
+    input_mapping_.erase(i);
+  }
+  CaffeMap<std::string, TensorShape> shape_hints_ordered;
+  for (const auto& kv : input_shape_hints) {
+    const auto it = input_reverse_mapping.find(kv.first);
+    if (it != input_reverse_mapping.end()) {
+      LOG(INFO) << "Adding input hint: " << it->second;
+      shape_hints_ordered.emplace(it->second, kv.second);
+    } else {
+      shape_hints_ordered.emplace(kv.first, kv.second);
+    }
+  }
+  return shape_hints_ordered;
+}
+
+void TensorRTTransformer::PruneUsedWeights(
+    Workspace* ws,
+    const NetDef& pred_net) {
+  std::unordered_set<std::string> used_weights;
+  for (const auto& op : pred_net.op()) {
+    for (const auto& i : op.input()) {
+      used_weights.emplace(i);
+    }
+  }
+
+  for (const auto kv: input_mapping_) {
+    // for weights that are not referenced anywhere, we remove it from the
+    // original workspace
+    if (!used_weights.count(kv.first)) {
+      VLOG(2) << "Removing unused weight blob: " << kv.second << " (" << kv.first << ")";
+      ws->RemoveBlob(kv.second);
+    }
+  }
+}
+
 // Cutting off the runnable part and replace with tensor ops. Asssume the nets
 // were topologically sorted
 void TensorRTTransformer::Transform(
     Workspace* ws,
     NetDef* pred_net,
     const std::unordered_map<std::string, TensorShape>& input_shape_hints) {
   CAFFE_ENFORCE(ws);
-
   auto shape_hints_ordered =
       SsaRewriteAndMapIO(ws, pred_net, input_shape_hints);
-  auto shape_hints = InferShapes(ws, pred_net, &shape_hints_ordered);
+  Workspace mapped_ws(ws, input_mapping_);
```

Comment:
Can input_mapping_ be a local variable instead of a class member?

Reply:
It can. And it used to be. But then in the ssa function I have to return multiple variables as a result. I made the trade-off here.